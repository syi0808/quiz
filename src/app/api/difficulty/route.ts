import { NextRequest, NextResponse } from 'next/server';
import cheerio from 'cheerio';

export async function GET(request: NextRequest) {
  const res = await fetch('https://opentdb.com/api_config.php');
  const html = await res.text();
  const $ = cheerio.load(html);

  const categories: { label: string; value: string }[] = $('select[name=trivia_difficulty]:first')
    .children()
    .toArray()
    .reduce((categories, optionElement) => {
      categories.push({
        label: (optionElement.children[0] as { data: string }).data,
        value: optionElement.attribs.value,
      });

      return categories;
    }, [] as { label: string; value: string }[]);

  return NextResponse.json(categories);
}
