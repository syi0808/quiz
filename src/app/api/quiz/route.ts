import { sleep } from '@/shared/utils/async';
import toCamelCase from '@/shared/utils/object';
import { NextRequest, NextResponse } from 'next/server';

export async function GET(request: NextRequest) {
  const { searchParams } = new URL(request.url);

  if (!searchParams.has('amount')) {
    searchParams.set('amount', '10');
  }

  if (searchParams.get('category') === 'any') {
    searchParams.delete('category');
  }

  if (searchParams.get('difficulty') === 'any') {
    searchParams.delete('difficulty');
  }

  const getResult = async () => {
    const res = await fetch(`https://opentdb.com/api.php?${searchParams.toString()}`);
    return await res.json();
  };

  do {
    const result = await getResult();

    if (result.response_code === 0) {
      return NextResponse.json(result.results.map(toCamelCase));
    } else {
      await sleep(100);
    }
  } while (1);
}
