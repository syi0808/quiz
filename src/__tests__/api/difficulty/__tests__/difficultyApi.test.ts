import { GET } from '../../../../app/api/difficulty/route';
import { NextRequest } from 'next/server';

describe('Get Difficulty API', () => {
  test('receive and parsing difficulties correctly from html', async () => {
    const response = await GET({
      url: 'http://localhost:3000/api/difficulty',
    } as NextRequest);

    const json = await response.json();

    expect(json).toMatchSnapshot();
  });
});
