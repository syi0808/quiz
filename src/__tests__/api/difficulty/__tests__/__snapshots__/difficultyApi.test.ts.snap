// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Difficulty API receive and parsing difficulties correctly from html 1`] = `
[
  {
    "label": "Any Difficulty",
    "value": "any",
  },
  {
    "label": "Easy",
    "value": "easy",
  },
  {
    "label": "Medium",
    "value": "medium",
  },
  {
    "label": "Hard",
    "value": "hard",
  },
]
`;
